### PR TITLE
Support post-quantum KYBER_LEVEL1 and P256_KYBER_LEVEL1 with FALCON_LEVEL1 in wolfMQTT.

### PR DESCRIPTION
**Pull request by anhu**

### CODE REVIEW

File: examples/mqttexample.c [hunk 1]
```diff
@@ -616,16 +632,52 @@ int mqtt_tls_cb(MqttClient* client)
...
-    if ((rc == WOLFSSL_SUCCESS) && (mTlsSniHostName != NULL)) {
-        rc = wolfSSL_CTX_UseSNI(client->tls.ctx, WOLFSSL_SNI_HOST_NAME,
-                mTlsSniHostName, (word16) XSTRLEN(mTlsSniHostName));
-        if (rc != WOLFSSL_SUCCESS) {
-            PRINTF("UseSNI failed");
+        /* REVIEWER: This has nothing to do with PQC but I think this fixes a
+         * a bug where we would have de-referenced client->tls.ctx even if it
+         * NULL.
+         * This comment should be removed during review.
+         */
```
[the remark this answers is not in the document]
Comment (dgarske) [reply examples/mqttexample.c:640]: Looks good. Remove your comment. You should put comments in the PR not in the code.

File: examples/mqttexample.c [hunk 2]
```diff
@@ -616,16 +632,52 @@ int mqtt_tls_cb(MqttClient* client)
...
 #endif /* HAVE_SNI */
+#ifdef HAVE_PQC
+        int group = 0;
+        if ((rc == WOLFSSL_SUCCESS) && (mTlsPQAlg != NULL)) {
+            if (XSTRCMP(mTlsPQAlg, "KYBER_LEVEL1") == 0) {
+                group = WOLFSSL_KYBER_LEVEL1;
+            } else if (XSTRCMP(mTlsPQAlg, "P256_KYBER_LEVEL1") == 0) {
+                group = WOLFSSL_P256_KYBER_LEVEL1;
+            } else {
+                PRINTF("invalid post-quantum KEM specified");
```
Comment (dgarske) [inline examples/mqttexample.c:657]: "Invalid" case.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfmqtt/mqtt_client.h [hunk 3]
```diff
@@ -482,6 +482,10 @@ WOLFMQTT_API int MqttClient_CancelMessage(
 
 /*! \brief      Performs network connect with TLS (if use_tls is non-zero value)
  *  Will perform the MqttTlsCb callback if use_tls is non-zero value
+ *  \note The callback function that cb points to does not have to initialize
```
[the remark this answers is not in the document]
Comment (anhu) [reply wolfmqtt/mqtt_client.h:485]: hmmm....I don't know what the difference is and I don't know why mqtt has its own custom ones in the first place.  I'll look into it tomorrow.

[the remark this answers is not in the document]
Comment (anhu) [reply wolfmqtt/mqtt_client.h:485]: OH !!!   You mean call those functions after the callback returns!!  I get it now. Will do tomorrow.

**Comment by anhu** [comment]
> Add a section in the README.md please showing steps to use PQ with MQTT. This would require the server to have PQ enabled to, so guessing there are many tips you can add.

Sure, will get to it tomorrow.

**Review by dgarske — APPROVED** [review]

**Comment by anhu** [comment]
Not ready for merge yet. Still need to do README.md additions.

**Review by dgarske — APPROVED** [review]

**Comment by anhu** [comment]
Firmware test is failing...repro'd on my local system. I still need to investigate.

**Comment by dgarske** [comment]
> 

Don't investigate, just retry the test. It isn't you... it is because our tests use a public server and the tests sometimes collide. We don't have our own MQTT broker for testing....

**Comment by anhu** [comment]
Already retried.  Repro'd locally, consistently and got seg fault.  Can't be result of collision .

**Comment by dgarske** [comment]
> Already retried. Repro'd locally, consistently and got seg fault. Can't be result of collision .

That's great. Should be easy then. I won't merge yet.

**Comment by anhu** [comment]
Yup.  Will investigate on Monday morning.

**Review by dgarske — APPROVED** [review]

**Pull request merged**